raft/server.cc

```diff
@@ -887,6 +874,10 @@ future<> server_impl::io_fiber(index_t last_stable) {
                     remove_from_rpc_config(addr);
                     _rpc->remove_server(addr.id);
                 }
+
+                if (!batch.configuration->is_joint() && _conf_commit_promise) {
+                    std::exchange(_conf_commit_promise, std::nullopt)->set_value();
```

Comment:
I'm afraid `batch.configuration` does not have to be committed yet. A configuration becomes part of a new `batch` as soon as it is appended - just look at the condition in `fsm::get_output()`:
```cpp
    if (_observed._last_conf_idx != _log.last_conf_idx() ||
            (_observed._current_term != _log.last_term() &&
             _observed._last_term != _log.last_term())) {
```
There's nothing about commit index here.

This `batch.configuration` was initially used to update RPC module - we need to be able to contact new members as soon as they are in the last "appended configuration".

Reply:
On the other hand, maybe we don't need to wait for the non-joint configuration to be committed?
Maybe it's enough to wait for the joint one?
In that case, this code would be enough - because if we get a non-joint config in the batch, it means that the previous joint config must have been committed.
This approach would require modifying the comments and adding more explanations. For example:
```cpp
    // Immediately, without yield, once the FSM discovers
    // this, it appends a non-joint entry.
    // Here we are waiting for this non-joint entry to be committed.
    co_await _conf_commit_promise.emplace().get_future();
```
This comment would have to explain something like "here we are waiting for the non-joint entry to appear; if it appears, it means that the joint entry was committed".

Reply:
We must wait since it makese set_configuration() api human. Unless we wait, a subsequent set_configuration() will throw an exception (can't append new config while in joint) which is not convenient.

Reply:
> There's nothing about commit index here.

yes, I missed it, will fix